### DISCUSSION TITLE
Combobox onSelect event or something equivalent

### DISCUSSION THREAD

**Discussion by gnapse** [ariakit/ariakit#2285] [Questions] [answered]
I want to use a combobox in a way similar to a dropdown select, but still keep it as a free-form input field. I even want the user to be able to type something not in the suggestions list.

So, I'm interested in a subset of the change events on the combobox. For instance, I do not want to know about every little change to the input field as the user types. But I want to know when the user leaves the field (`onBlur`) which works. And I also want to know when the user selected something from the suggested options.

That last bit is what I'm missing. Is there a way to get notified whenever the user selected a suggested value? As silly as the question sounds (because I expected to find this right on the surface somewhere in the API), I have not found it yet. I inspected all the type definitions for the combobox state, the state props, the component props, nothing. I tried hooking up to the `onChange` event, nothing.

Hoping someone can help with this. Thanks!

**Comment by gnapse** [comment, marked as answer]
I think I found an answer from [this example](https://ariakit.org/examples/combobox-textarea): using `onClick` in a `ComboboxItem`.

**Reply by diegohaz** [reply]
Yeah, we try to use native events as most as possible.

**Reply by IanVS** [reply]
Native `<select>` elements do emit a `change` event: https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/change_event.  Without it, I'm struggling to hook up an Ariakit Select component to react-hook-form: https://react-hook-form.com/docs/useform/register, since I need to wire up an `onChange` callback, which takes as its argument the change event with a target, not an updated value.  

As a workaround, I can directly set the value with `setValue` in the onClick callbacks.  Have to be careful to use `shouldValidate: true` though.  And hopefully there aren't other gotchas...

Edit: there was another gotcha.  Changing the value by focusing the field and pressing a key on the keyboard was not registered.  Rendering a custom hidden select seems to be the workaround to handle that.

**Comment by ellemedit** [comment]
I guess ariakit Select should expose inner native select in implementation.
I wish select works as native select for onChange, value, defaultValue and name attribute.
I tried to mimic native select with ariakit Select by myself but ariakit Select hijack the name prop and use itself. it is hard to integrate with other library such as react-hook-form!

please see this issue seriously. I have to do hack to solve this. @diegohaz

**Reply by diegohaz** [reply]
That won't work. When you select an option on the custom Select component, the `onChange` event is not dispatched on the native select element. You should use the controlled version of react-hook-form instead.

But you can render a native select yourself if you want: https://github.com/ariakit/ariakit/issues/2491#issuecomment-1571144023

**Reply by ellemedit** [reply]
thank you!

```jsx
function HiddenSelect({ store, ...props }: HiddenSelectProps) {
  const value = store.useState("value");
  const items = store.useState("items");
  const values = items.map((item) => item.value);
  const ref = useRef();

  useEffect(() => {
    ref.current.dispatchEvent(new Event('change', { bubbles: true }));
  }, [value])

  return (
    <select ref={ref} hidden value={value} onChange={(event) => {
      console.log("###", event.currentTarget.selectedOptions);
    }} {...props}>
      {!values.includes(value) && <option value={value}>{value}</option>}
      {values.map((value) => (
        <option key={value} value={value}>
          {value}
        </option>
      ))}
    </select>
  );
}
```

this looks working well as workaround.
